fedimint-client/src/oplog.rs

```diff
@@ -64,45 +67,107 @@ impl OperationLog {
         limit: usize,
         start_after: Option<ChronologicalOperationLogKey>,
     ) -> Vec<(ChronologicalOperationLogKey, OperationLogEntry)> {
+        const EPOCH_DURATION: Duration = Duration::from_secs(60 * 60 * 24 * 7);
+
         let mut dbtx = self.db.begin_transaction_nc().await;
-        let operations: Vec<ChronologicalOperationLogKey> = dbtx
-            .find_by_prefix_sorted_descending(&ChronologicalOperationLogKeyPrefix)
-            .await
-            .map(|(key, ())| key)
-            // FIXME: this is a schlemil-the-painter algorithm that will take longer the further
-            // back in history one goes. To avoid that I see two options:
-            //   1. Add a reference to the previous operation to each operation log entry,
-            //      essentially creating a linked list, which seem a little bit inelegant.
-            //   2. Add an option to prefix queries that allows to specify a start key
-            //
-            // The current implementation may also skip operations due to `SystemTime` not being
-            // guaranteed to be monotonous. The linked list approach would also fix that.
-            .skip_while(|key| {
-                let skip = if let Some(start_after) = start_after {
-                    key.creation_time >= start_after.creation_time
-                } else {
-                    false
-                };
-
-                std::future::ready(skip)
-            })
-            .take(limit)
-            .collect::<Vec<_>>()
-            .await;
+        let operation_log_keys = if let Some(start_after) = start_after {
+            let Some(last_key) = dbtx
+                .find_by_prefix(&ChronologicalOperationLogKeyPrefix)
+                .await
+                .map(|(key, ())| key)
+                .next()
+                .await
+            else {
+                return vec![];
+            };
+
+            // We want to fetch all operations that were created before `start_after`, going
+            // backwards in time. This means "start" generally means a later time than
+            // "end". Only when creating a rust Range we have to swap the terminology (see
+            // comment there).
+            let epochs_rev_ranges = (0..)
+                .map(|epoch| start_after.creation_time - epoch * EPOCH_DURATION)
```

Comment:
Are we ok with just not expecting people to run our app in 1970?

Reply:
[They weren't ready for it anyway](https://github.com/user-attachments/assets/c79e271c-64f0-4125-8d05-485e9fe14e43)